Ia32 structs

P-E-P opened the pull request:
Flatten structs and use more meaningful names.

P-E-P committed:
feat(cargo): Add bit_field dependency

Add a dependency to the bit_field crate.

P-E-P committed:
refactor(ia32): Refactor descriptor segment

Refactor descriptor segment to split internal fields.

P-E-P committed:
refactor: Reformat code

Reformat the whole code.

P-E-P committed:
refactor(ia32): Refactor segment descriptor module

Split segment descriptor module with two new submodules, one for every
sub structure. This will allow shorter and cleaner code files.

P-E-P committed:
refactor(ia32/segment): Remove old lower and upper

Remove old lower/upper structures for segment descriptors.

P-E-P committed:
feat(ia32/segment): Implement permissions display

Implement the display trait for struct Permissions.

P-E-P committed:
feat(ia32/segment): Add permission display

Add implementation to display permissions of a segment.

P-E-P committed:
feat(ia32/segment): Add configuration display

Implement display for configuration structure.

P-E-P committed:
feat(segment): Add segment display

Add implementation of Display for segment descriptors.

P-E-P committed:
feat(ia32/tss): Add configuration structure

Add the configuration bitfield for the tss descriptor.

P-E-P committed:
feat(ia32/tss): Add tss permissions descriptor

Add the permission bitfield structure for the tss descriptor.

P-E-P committed:
feat(ia32{e}/tss): Use the new internal structures

Use the new permissions and configuration internal structures for tss
descriptors.

P-E-P committed:
refactor(ia32/tss): Remove upper structure

The upper structure for tss descriptor is not used anymore, hence it's
removal.

P-E-P committed:
refactor(interrupt gate): Change format

Change the internal format of interrupt gates to multiple fields instead
of a lower and upper generic field.

P-E-P committed:
refactor(ia32/segment): Change builder pattern

The old builder pattern was using mutable references. While developping
tss segment we encountered references on unaligned struct members
because of function packing. This made the builder pattern with
references much less attractive hence this change. The clone call should
not have any meaningful runtime overhead since all those structures
implements the Copy trait.

P-E-P committed:
refactor(ia32{e}/tss): Change builder pattern

Change the builder pattern of tss structures to remove the mutable
references.

P-E-P committed:
refactor(ia32): Reformat code

Apply reformat script.

P-E-P committed:
refactor(ia32/trap): Refactor trap desc structure

Refactor the trap descriptor structure, it does no longer use the upper
and lower opaque fields but rather new meaningful fields instead.

P-E-P committed:
refactor(ia32): Remove old upper structure

Remove old unused upper structure modules for interrupts and trap
descriptors.

P-E-P committed:
fix(ia32): Set interrupt descriptor to present

When constructing a descriptor we give it value, hence it should be
marked as present by default since we're actively changing the default
values.

P-E-P committed:
refactor(ia32/task): Change task desc structure

Change task descriptors internal fields structure by removing the old
upper and lower fields and replacing them with more meaningful field
name with appropriate size.

P-E-P committed:
refactor(ia32/trap): Reformat code

Apply reformat script.

P-E-P committed:
refactor(ia32/task): Remove old lower and upper

Remove old unused lower and upper modules for the task descriptor
structure.

P-E-P committed:
refactor(ia32/gate): Remove unused gate macro

Since we're using an unsafe transmute now and since lower and upper
fields do not exist anymore this macro is useless and should be deleted.

P-E-P committed:
refactor(ia32e/gate): Remove old lower and upper

Remove old lower and upper fields to flatten the structure.

P-E-P committed:
refactor(ia32/gate): Remove unused lower module

This module was lastly used by the ia32e gate structure. Since recent
changes it is not required anymore.

P-E-P committed:
refactor(ia32e): Reformat code

Apply reformat script.

P-E-P committed:
refactor(ia32e/gate): Remove unused upper module

Remove unused upper module since recent changes removing those fields.

P-E-P committed:
refactor(ia32/segment): Set present

Set segment created from constructors (hence containing data) to
present by default.

P-E-P committed:
refactor(ia32/tss): Set present

Set tss descriptor present by default when created from a constructor.

P-E-P committed:
refactor(macros): Remove now unused macros

Offset and setbit macro are not used anymore since we do not rely on
const flags to set values.

P-E-P committed:
refactor(segment): Add unit tests for config

Add unit tests for ia32 segment descriptor configuration structure.

P-E-P committed:
doc(segment): Document configuration field

Document the configuration field as well as it's methods.

P-E-P committed:
doc(ia32/segment): Document SegmentDescriptor

Document the SegmentDescriptor structure and it's methods.

P-E-P committed:
refactor(descriptor): Remove unused functions

Split limit and split address functions are not in use anymore thanks to
bitfields, hence the removal.

P-E-P committed:
doc(segment): Add doc to configuration offsets

Add more documentation to segment configuration field offset values.

P-E-P committed:
doc(ia32/segment): Document segment descriptors

Document even more the segment descriptor structure.

P-E-P committed:
test(segment): Test for structure size

This structure should have a well defined size and it should not be
changed, at least not easily. That's why this commit introduces a unit
test to check for the configuration's type size.

P-E-P committed:
refactor(ia32/segment): Change offsets

Change the permission offsets formats to submodules instead of prefixes.

P-E-P committed:
doc(segment/permissions): Add documentation

Add documentation to segments permission field.

P-E-P committed:
fix(ia32/segment): Fix SegmentType conversion

Fix SegmentType conversion to binary. Accessed and execute/conforming
were inverted.

P-E-P committed:
test(segment/permissions): Add SegmentType tests

Add some tests for segment type setter.

P-E-P committed:
test(ia32/segment): Add unit tests for setters

Add multiple unit tests for internal field setters.

P-E-P committed:
refactor(ia32/segment): Refactor format usage

Reduce the overall line length in the format function by adding a use
statement.

P-E-P committed:
doc(ia32/granularity): Document granularity

Document granularity enum members.

P-E-P committed:
refactor(gate): Improve gate display format

Improve gate display format by collecting usefull known fields.

P-E-P committed:
doc(gate): Add generic gate documentation

Add documentation for the generic gate structure.

P-E-P committed:
refactor(gate): Apply format script

Reformat the code by applying reformat script.

P-E-P committed:
refactor(interrupt): Change configuration offsets

Change configuration's offset formats from global with prefixes/suffixes
to a more organized structure with submodules.

P-E-P committed:
doc(interrupt): Add configuration offset doc

Add the documentation for the interrupt configuration's offset values.

P-E-P committed:
doc(ia32/interrupt): Add configuration doc

Add documentation for interrupt gates configuration structure.

P-E-P committed:
fix(ia32/segment): Fix unit test name

One unit test had a name similar to another one.

P-E-P committed:
test(ia32/interrupt): Add unit tests

Add more unit tests for interrupt gates configuration structure.

P-E-P committed:
doc(ia32/interrupt): Fix documentation

Fix documentation about configuration structure offset in interrupt
gates.

P-E-P committed:
test(ia32/interrupt): Add zeroes field unit test

Add a unit test to assert that the field of zeroes are effectively
zeroed.

P-E-P committed:
refactor(ia32/task): Refactor task configuration

Add documentation and unit tests to a task gate configuration field.

P-E-P committed:
refactor(ia32/trap): Refactor trap configuration

Add documentation to trap configuration structure as well as some unit
tests.

P-E-P committed:
refactor(ia32e/gate): Change configuration offsets

Refactor configuration offsets in ia32e gate descriptor to remove common
prefixes and suffixes.

P-E-P committed:
refactor(ia32e/gate): Refactor gate configuration

Add some unit tests and some documentation to the configuration
structure of ia32e gate.

P-E-P committed:
fix(ia32/trap): Fix documentation

The link to the parent structure was incorrect.

P-E-P committed:
doc(ia32e/gate): Add panic warning to ist doc

Add a panic warning to ist field documentation for the configuration
structure.

P-E-P committed:
refactor(ia32e/gate): Remove unused import

P-E-P committed:
refactor(ia32/tss): Refactor configuration

Add documentation and unit tests to tss configuration structure.

P-E-P committed:
refactor(ia32/interrupt): Refactor InterruptGate

Add documentation to interrupt gate and some unit tests.

P-E-P committed:
refactor(ia32/segment): Add unit tests

Add some unit tests to ia32 SegmentDescriptor structure.

P-E-P committed:
refactor(ia32/task): Add documentation and tests

Add some documentation and tests to the TaskGate structure.

P-E-P committed:
doc(ia32/interrupt): Change link

Change link to InterruptGate.

P-E-P committed:
refactor(ia32/trap): Add documentation and tests

Add documentation and more tests to TrapGate structure.

P-E-P committed:
refactor(ia32/tss): Refactor permissions

Add documentation and unit tests to Permissions structure of the TSS
descriptor.

P-E-P committed:
refactor(ia32{e}/tss): Add documentation and tests

Add unit tests to TssDescriptor structures and some documentation.

P-E-P committed:
refactor(ia32/address): Reformat code

Apply reformat script.

alexandrefresnais commented on src/arch/ia32/descriptor/gate/interrupt/configuration.rs:
```diff
@@ -0,0 +1,150 @@
+
+impl Default for Configuration {
+    fn default() -> Self {
+        Configuration(0x0600)
```
Maybe a comment describing in details what the default configuration is, could help

P-E-P replied:
Done in ba5c2c8

alexandrefresnais commented on src/arch/ia32/descriptor/gate/task/configuration.rs:
```diff
@@ -0,0 +1,111 @@
+
+impl Default for Configuration {
+    fn default() -> Self {
+        Configuration(0x0500)
```
Here too, we could give details

P-E-P replied:
Done in ba5c2c8

alexandrefresnais commented on src/arch/ia32/descriptor/gate/trap.rs:
```diff
@@ -1,41 +1,91 @@
         }
     }
 
+    /// Change a [``]'s privilege level by another one.
```
It think the `TrapGate` word is missing between the brackets.

P-E-P replied:
Done in 801000d

alexandrefresnais commented on src/arch/ia32/descriptor/gate/trap/configuration.rs:
```diff
@@ -0,0 +1,150 @@
+
+impl Default for Configuration {
+    fn default() -> Self {
+        Configuration(0x0700)
```
Here again, if you agree with specifying the bits set

P-E-P replied:
Done in ba5c2c8

alexandrefresnais commented on src/arch/ia32/descriptor/gate/trap/configuration.rs:
```diff
@@ -0,0 +1,150 @@
+        use offset::{privilege_level, PRESENT, SIZE};
+        write!(
+            f,
+            "Size: {}\nDPL: {}\nPresent {}",
```
Here, and in every other `Configuration` display traits, the semicolon `:`  is missing in after "Present"

P-E-P replied:
Done in 57418d7

alexandrefresnais commented on src/arch/ia32/descriptor/segment.rs:
```diff
@@ -1,155 +1,301 @@
 
+/// The default operation size modes of a memory segment.
+///
+/// It could be either 16 bits (`0`) or 32 bits (`0`). It performs different
```
I believe it was supposed to be `or 32 bits ('1')`

P-E-P replied:
Fixed in 141a5a9

alexandrefresnais commented on src/arch/ia32/descriptor/segment.rs:
```diff
@@ -1,155 +1,301 @@
+/// The default operation size modes of a memory segment.
+///
+/// It could be either 16 bits (`0`) or 32 bits (`0`). It performs different
+/// functions depending on wheter the segment descriptor is an executable code
```
`whether`

P-E-P replied:
Fixed in 141a5a9

alexandrefresnais commented on src/arch/ia32/descriptor/tss/permissions.rs:
```diff
@@ -0,0 +1,139 @@
+
+impl Default for Permissions {
+    fn default() -> Self {
+        Permissions(0b00001001)
```
Here again, do you want to precise what is set ?

P-E-P replied:
Done in ce93a17

alexandrefresnais commented on src/arch/ia32/descriptor/tss/permissions.rs:
```diff
@@ -0,0 +1,139 @@
+        use offset::{privilege_level, BUSY, PRESENT};
+        write!(
+            f,
+            "Busy: {}\nDPL: {}\nPresent {}",
```
Here too if you want to put a semicolon after "Present"

P-E-P replied:
Done in 57418d7

alexandrefresnais commented on src/arch/ia32e/descriptor/gate.rs:
```diff
@@ -1,92 +1,83 @@
-    ///  Segment selector bits (32:64).
-    upper: Upper,
-    /// Procedure's entry point offset high bits.
+    offset_15_0: u16,
```
Don't you want to comment the role of every field as you did for other structs ?

P-E-P replied:
Fixed in fe6ea60

alexandrefresnais commented on src/arch/ia32/descriptor/tss.rs:
```diff
@@ -1,65 +1,139 @@
-#[derive(Copy, Clone)]
+#[derive(Default, Copy, Clone)]
 #[repr(C, packed)]
 pub struct TssDescriptor {
```
Don't you want to implement `fmt::Display` for this structure ?
You can leave a `todo!` if you want as in the ia32e gate.rs.

P-E-P replied:
The current display format is not the final one, I'm not too sure about what we should do or what we should need. We'll keep the current code and implement proper display trait for those  structure in another PR.

alexandrefresnais commented on src/arch/ia32e/descriptor/tss.rs:
```diff
@@ -1,59 +1,123 @@
     reserved: u32,
 }
 
 impl TssDescriptor {
```
No `fmt::display` implementation here too ?

P-E-P replied:
The current display format is not the final one, I'm not too sure about what we should do or what we should need. We'll keep the current code and implement proper display trait for those  structure in another PR.

alexandrefresnais reviewed (changes requested):
Mad impressive.
I left some minor comments.
Sorry and thank you for correcting the coding style in `address.rs`.

P-E-P committed:
doc(ia32/gates): Document default bits

Default bits in the configuration structure for interrupt/trap/task
gates were not documented.

P-E-P committed:
doc(ia32/trap): Fix missing link in documentation

A link syntax was left empty.

P-E-P committed:
refactor(ia32): Fix missing semicolon in Display

Add missing semicolon in Display trait.

P-E-P committed:
doc(tss): Add documentation to tss specific bits

Add some documentation to the tss configuration structure about the
default bits.

P-E-P committed:
doc(ia32/segment): Fix documentation

Some informations in the documentation where wrong along some typos.

P-E-P committed:
doc(ia32e/gate): Document structure and methods

Document the Gate descriptor for ia32e.

alexandrefresnais reviewed (approved):
LGTM, very well done!
